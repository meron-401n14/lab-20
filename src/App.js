import React from "react";


import Input from "./Components/Contactinput.js";
import Select from "./Components/Select.js";
import CheckBox from "./Components/CheckBox.js";
import Button from "./Components/Button.js";
import "./styles/styles.scss";


/**
 
 *this is a parent class which manage state pass props
 * @class MangeContact
 * @constructor   Props pass data thus will not be changed 
 * state pass  used for data that will change (name, phone ...)
 
*/
 

class ManageContact extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      newContact: {
        name: "",
        phone: "",
        gender: "",
        positions: []
      },
      genderOptions: ["Male", "Female", "others"],
      positionOptions: ["Instructor ", "TA", "Student"]
    };

 // here we bind all handle function of components 
    this.handleTextArea = this.handleTextArea.bind(this);
    this.handleFullName = this.handleFullName.bind(this);
    this.handlePhone = this.handlePhone.bind(this);
    this.handleInput = this.handleInput.bind(this);
    this.handleCheckBox = this.handleCheckBox.bind(this);
    this.handleClearForm = this.handleClearForm.bind(this);
  }
/**
 * this function set a state to the current 
 * @param {object} e 
 * this function handles update input name 
 * use prevState to be acurate on the update 
 */
  handleFullName(e) {
    let value = e.target.value;
    this.setState(
      prevState => ({
        newContact: {
          ...prevState.newContact,
          name: value
        }
      }),
      () => console.log(this.state.newContact)
    );
  }

  /**
 * this function set a state to the current 
 * @param {event} e 
 * this function handles update phone input value 
 * use prevState to be acurate on the update 
 */

  handlePhone(e) {
    let value = e.target.value;
    this.setState(
      prevState => ({
        newContact: {
          ...prevState.newContact,
          phone: value
        }
      }),
      () => console.log(this.state.newContact)
    );
  }

    /**
 * this function set a state to the current 
 * @param {event} e 
 * this function handles update of other input values   
 * use prevState to be acurate on the update 
 */
  handleInput(e) {
    let value = e.target.value;
    let name = e.target.name;
    this.setState(
      prevState => ({
        newContact: {
          ...prevState.newContact,
          [name]: value
        }
      }),
      () => console.log(this.state.newContact)
    );
  }

  /**
   * updates text area placeholder and other values 
   * @param {event} e 
   */
  handleTextArea(e) {
    console.log("Inside handleTextArea");

    this.setState(
      prevState => ({
        newContact: {
          ...prevState.newContact
        }
      }),
      () => console.log(this.state.newContact)
    );
  }

  handleCheckBox(e) {
    const newSelection = e.target.value;
    let newSelectionArray;
    if (this.state.newContact.positions.indexOf(newSelection) > -1) {
      newSelectionArray = this.state.newContact.positions.filter(
        s => s !== newSelection
      );
    } else {
      newSelectionArray = [...this.state.newContact.positions, newSelection];
    }
    this.setState(prevState => ({
      newContact: { ...prevState.newContact, positions: newSelectionArray }
    }));
  }
/**
 * this function set update state of details 
 * @param {event} e 
 * function for Button click 
 */
  handleClearForm(e) {
    e.preventDefault();
    this.setState({
      newContact: {
        name: "",
        phone: "",
        gender: "",
        positions: []
      }
    });
  }

  render() {
    return (
      <div className="col-md-6">
        <div className="contactMgr">
          <form className="container-fluid">
            <Input
              id={'name'}
              inputtype={"text"}
              title={"Full Name"}
              name={"name"}
              value={this.state.newContact.name}
              placeholder={"Enter your name"}
              handleChange={this.handleInput}
            />{" "}
            {/* Name of the person */}
            <Input
              id={'phone'}
              inputType={"text"}
              name={"phone"}
              title={"Phone"}
              value={this.state.newContact.phone}
              placeholder={"Enter your phone"}
              handleChange={this.handlePhone}
            />{" "}
            <Select
              title={"Gender"}
              name={"gender"}
              options={this.state.genderOptions}
              value={this.state.newContact.gender}
              placeholder={"Select Gender"}
              handleChange={this.handleInput}
            />
            <CheckBox
              title={"Position"}
              name={"positions"}
              options={this.state.positionOptions}
              selectedOptions={this.state.newContact.positions}
              handleChange={this.handleCheckBox}
            />
          </form>
        </div>
        <div className="Display">
          <p />
          <h3>Contact Information</h3>
          <li>{this.state.newContact.name}</li>
          <li>{this.state.newContact.phone}</li>
          <li>{this.state.newContact.gender}</li>
          <li>{this.state.newContact.positions}</li>
        </div>
        <Button
          action={this.handleClearForm}
          type={"primary"}
          title={"Clear"}
          style={buttonStyle}
        />
      </div>
    );
  }
}

const buttonStyle = {
  margin: "10px 10px 300px 100px"
};

export default ManageContact;
