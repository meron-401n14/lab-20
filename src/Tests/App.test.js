import React from "react";

import renderer from 'react-test-renderer';
import { create } from "react-test-renderer"
import { configure, render, mount, shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import App from "../App";
import input from "../Components/Contactinput";


configure({ adapter: new Adapter() });


describe("APP test render", () => {

  it('renders without crashing', () => {
    shallow(<App />)
  })

  it('This tests porps for contactinput component passed and exist should have input for name and phone', () => {
    const wrapper = shallow(<App />)
    expect(wrapper.find('Input#name')).toHaveLength(1);
    expect(wrapper.find('Input#phone')).toHaveLength(1);
  });


  it('Contact Information', () => {
    const wrapper = shallow(<App />)
    const Information = <h3>Contact Information</h3>

    expect(wrapper.contains(Information)).toEqual(true)
  })

  it("renders ContactInput", () => {
    const fullNameInputComponent = renderer.create(<form />).toJSON();
    expect(fullNameInputComponent).toMatchSnapshot();
  });


  it("Should have a button component", () => {

    function Button(props) {
      return <button>Clear Form</button>;
    }
    const button = create(<Button />);
    expect(button.toJSON()).toMatchSnapshot();
  });


  it('accepts contact props', () => {
    const contact = {

      name: 'Meron Si',
      phone: '23456723',
      Gender: 'female',
      position: 'student'

    }
    const wrapper = mount(<input contact={contact} />);
    expect(wrapper.props().contact).toEqual(contact)
  })



});

describe('Displays form elements and a state printer', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = mount(<App />);
  });
  const originalError = console.error;
  console.error = jest.fn();
  it('should render one <h3>', () => {
    expect(wrapper.find('h3')).toHaveLength(1);
    console.error = originalError;
  });
  it('should render one <Form>', () => {

    expect(wrapper.find('p')).toHaveLength(1);
    console.error = originalError;
  });
  it('should render 1 <label>s', () => {

    expect(wrapper.find('label')).toHaveLength(7);
    console.error = originalError;
  })

});




// done 
