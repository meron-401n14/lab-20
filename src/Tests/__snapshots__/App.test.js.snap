// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`APP test render Should have a button component 1`] = `
<button>
  Clear Form
</button>
`;

exports[`APP test render renders ContactInput 1`] = `<form />`;
